pybop/models/base_model.py

```diff
@@ -134,32 +153,43 @@ def build(
 
         self.n_states = self._built_model.len_rhs_and_alg  # len_rhs + len_alg
 
-    def set_init_soc(self, init_soc: float):
+    def set_initial_state(self, initial_state: Optional[float] = None):
         """
-        Set the initial state of charge for the battery model.
+        Set the initial concentrations for the battery model.
 
         Parameters
         ----------
-        init_soc : float
-            The initial state of charge to be used in the model.
-        """
-        if self._built_initial_soc != init_soc:
-            # reset
-            self._model_with_set_params = None
-            self._built_model = None
-            self.op_conds_to_built_models = None
-            self.op_conds_to_built_solvers = None
-
-        param = self.pybamm_model.param
-        self._parameter_set = (
-            self._unprocessed_parameter_set.set_initial_stoichiometries(
-                init_soc, param=param, inplace=False
-            )
-        )
-        # Save solved initial SOC in case we need to rebuild the model
-        self._built_initial_soc = init_soc
+        initial_state : float or str, optional
+            If float, this value is used as the initial state of charge (as a decimal between 0
+            and 1). If str ending in "V", this value is used as the initial open-circuit voltage.
+            Defaults to None, indicating that the existing initial concentrations will be used.
+        """
+        if "Initial SoC" in self._unprocessed_parameter_set.keys():
+            initial_soc = self.get_initial_state(initial_state)
+            self._unprocessed_parameter_set.update({"Initial SoC": initial_soc})
 
-    def set_params(self, rebuild=False):
+        else:
+            # Point pybamm variables at pybop variables
+            self.model = self.pybamm_model
+            self._model = self.pybamm_model
+            self._unprocessed_parameter_values = self._unprocessed_parameter_set
+
+            pybamm_set_initial_state = pybamm.Simulation.set_initial_soc
+            pybamm_set_initial_state(self, initial_state, inputs=None)
+
+            # Update the default parameter set for consistency
+            self._unprocessed_parameter_set = self._parameter_values
+
+            # Clear the pybamm variables
+            del self.model
+            del self._model
+            del self._unprocessed_parameter_values
+            del self._parameter_values
+
+        # Use a copy of the updated default parameter set
+        self.parameter_set = self._unprocessed_parameter_set
```

Comment:
If we want to be explicit about copying, we could use (without the setter):
`self._parameter_set = self._unprocessed_parameter_set.copy()`
Otherwise we are copying twice?

Reply:
I have resolved this by removing the parameter_set.setter, which was only used internally.